File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/python/exercises/practice/rest-api/rest_api.py
import json
from collections import defaultdict

class RestAPI:
    class User:
        def __init__(self, name):
            self.name = name
            self.owes = defaultdict(float)
            self.owed_by = defaultdict(float)
        
        @property
        def balance(self):
            return sum(self.owed_by.values()) - sum(self.owes.values())
        
        def to_json(self):
            return {
                "name": self.name,
                "owes": {k: round(v, 2) for k, v in self.owes.items() if v > 0},
                "owed_by": {k: round(v, 2) for k, v in self.owed_by.items() if v > 0},
                "balance": round(self.balance, 2)
            }
    
    def __init__(self, database=None):
        self.users = {}
        if database:
            for user_data in database.get('users', []):
                user = self.User(user_data['name'])
                # Initialize owes relationships
                for borrower, amount in user_data['owes'].items():
                    user.owes[borrower] = amount
                # Initialize owed_by relationships
                for lender, amount in user_data['owed_by'].items():
                    user.owed_by[lender] = amount
                self.users[user.name] = user
    
    def get(self, url, payload=None):
        if url == '/users':
            if payload:
                requested_users = json.loads(payload).get('users', [])
                filtered = [u for name, u in self.users.items() if name in requested_users]
            else:
                filtered = list(self.users.values())
            
            filtered.sort(key=lambda u: u.name)
            return json.dumps({"users": [u.to_json() for u in filtered]})
        
        return json.dumps({})
    
    def post(self, url, payload=None):
        if url == '/add':
            data = json.loads(payload) if payload else {}
            name = data.get('user', '')
            if name in self.users:
                return json.dumps({"error": "User already exists"}), 400
            
            new_user = self.User(name)
            self.users[name] = new_user
            return json.dumps(new_user.to_json())
        
        elif url == '/iou':
            data = json.loads(payload) if payload else {}
            lender_name = data.get('lender', '')
            borrower_name = data.get('borrower', '')
            amount = data.get('amount', 0.0)
            
            lender = self.users.get(lender_name)
            borrower = self.users.get(borrower_name)
            
            if not lender or not borrower:
                return json.dumps({"error": "User not found"}), 404
            
            # Update lender's owed_by and borrower's owes
            lender.owed_by[borrower_name] += amount
            borrower.owes[lender_name] += amount
            
            # Simplify mutual debts
            for user1, user2 in [(lender, borrower), (borrower, lender)]:
                debt = min(user1.owed_by.get(user2.name, 0), user2.owed_by.get(user1.name, 0))
                if debt > 0:
                    user1.owed_by[user2.name] -= debt
                    user2.owed_by[user1.name] -= debt
                    if user1.owed_by[user2.name] <= 0:
                        del user1.owed_by[user2.name]
                    if user2.owed_by[user1.name] <= 0:
                        del user2.owed_by[user1.name]
            
            # Return updated users sorted by name
            updated_users = sorted([lender, borrower], key=lambda u: u.name)
            return json.dumps({"users": [u.to_json() for u in updated_users]})
        
        return json.dumps({})

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/javascript/exercises/practice/rest-api/rest-api.js
//
// This is only a SKELETON file for the 'Rest API' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class RestAPI {
  constructor(database = { users: {} }) {
    this.db = database;
    if (!this.db.users) {
      this.db.users = {};
    }
  }

  get(url) {
    if (url.startsWith('/users')) {
      const queryPart = url.split('?')[1] || '';
      const params = new URLSearchParams(queryPart);
      const requestedNames = params.get('users') ? params.get('users').split(',') : [];
      
      const users = requestedNames.length 
        ? requestedNames.map(name => this.db.users[name]).filter(user => user !== undefined)
        : Object.values(this.db.users);
      
      const processedUsers = users.map(user => {
        const owedByTotal = Object.values(user.owed_by || {}).reduce((sum, val) => sum + val, 0);
        const owesTotal = Object.values(user.owes || {}).reduce((sum, val) => sum + val, 0);
        const balance = owedByTotal - owesTotal;
        
        return {
          name: user.name,
          owes: user.owes || {},
          owed_by: user.owed_by || {},
          balance: parseFloat(balance.toFixed(2))
        };
      }).sort((a, b) => a.name.localeCompare(b.name));
      
      return { users: processedUsers };
    }
    throw new Error(`Invalid endpoint: ${url}`);
  }

  post(url, payload) {
    if (url === '/add') {
      const userName = payload.user;
      if (this.db.users[userName]) {
        throw new Error('User already exists');
      }
      
      const newUser = {
        name: userName,
        owes: {},
        owed_by: {},
        balance: 0.0
      };
      
      this.db.users[userName] = newUser;
      return newUser;
    }
    
    if (url === '/iou') {
      const { lender, borrower, amount } = payload;
      if (!this.db.users[lender] || !this.db.users[borrower]) {
        throw new Error('User not found');
      }
      
      // Update lender's owed_by and borrower's owes
      const lenderUser = this.db.users[lender];
      const borrowerUser = this.db.users[borrower];
      
      // Calculate net debt
      const existingDebtFromBorrower = lenderUser.owed_by[borrower] || 0;
      const existingDebtToLender = borrowerUser.owes[lender] || 0;
      
      // Calculate new debt
      let newLenderDebt = existingDebtFromBorrower - existingDebtToLender + amount;
      let newBorrowerDebt = existingDebtToLender - existingDebtFromBorrower - amount;
      
      // Update lender's records
      if (newLenderDebt > 0) {
        lenderUser.owed_by[borrower] = newLenderDebt;
        delete lenderUser.owes[borrower];
      } else if (newLenderDebt < 0) {
        lenderUser.owes[borrower] = -newLenderDebt;
        delete lenderUser.owed_by[borrower];
      } else {
        delete lenderUser.owed_by[borrower];
        delete lenderUser.owes[borrower];
      }
      
      // Update borrower's records
      if (newBorrowerDebt > 0) {
        borrowerUser.owes[lender] = newBorrowerDebt;
        delete borrowerUser.owed_by[lender];
      } else if (newBorrowerDebt < 0) {
        borrowerUser.owed_by[lender] = -newBorrowerDebt;
        delete borrowerUser.owes[lender];
      } else {
        delete borrowerUser.owed_by[lender];
        delete borrowerUser.owes[lender];
      }
      
      // Prepare response with updated users
      const responseUsers = [lenderUser, borrowerUser].map(user => {
        const owedTotal = Object.values(user.owed_by || {}).reduce((sum, val) => sum + val, 0);
        const owesTotal = Object.values(user.owes || {}).reduce((sum, val) => sum + val, 0);
        return {
          name: user.name,
          owes: { ...user.owes },
          owed_by: { ...user.owed_by },
          balance: parseFloat((owedTotal - owesTotal).toFixed(2))
        };
      }).sort((a, b) => a.name.localeCompare(b.name));
      
      return { users: responseUsers };
    }
    
    throw new Error(`Invalid endpoint: ${url}`);
  }
}
